pkg/agent/dial.go

```diff
@@ -176,6 +144,49 @@ func connect(logger *logging.Logger, transport Transport, mode, prompter string,
 	return stream, false, false, nil
 }
 
+// agentInvocationPath computes the agent invocation path, relative to the user's home
+// directory on the remote.
+func agentInvocationPath(cmdExe bool) string {
+	// Unless we have reason to assume that this is a cmd.exe environment, we
+	// construct a path using forward slashes. This will work for all POSIX
+	// systems and POSIX-like environments on Windows. If we know we're hitting
+	// a cmd.exe environment, then we use backslashes, otherwise the invocation
+	// won't work. Watching for cmd.exe to fail on commands with forward slashes
+	// is actually the way that we detect cmd.exe environments.
+	//
+	// HACK: We're assuming that none of these path components have spaces in
+	// them, but since we control all of them, this is probably okay.
+	//
+	// HACK: When invoking on Windows systems (whether inside a POSIX
+	// environment or cmd.exe), we can leave the "exe" suffix off the target
+	// name. Fortunately this allows us to also avoid having to try the
+	// combination of forward slashes + ".exe" for Windows POSIX environments.
+	//
+	// HACK: When invoking on cmd.exe, we leave off the ~ prefix, since cmd.exe
+	// doesn't recognize it. In most cases the initial working directory for SSH
+	// commands is the home directory, but when possible we try to be explicit,
+	// to work around systems that use a different directory, such as Coder
+	// workspaces, which allow different initial working directories to be
+	// configured.
+	pathSeparator := "/"
+	pathComponents := []string{filesystem.HomeDirectorySpecial}
+	if cmdExe {
+		pathSeparator = "\\"
+		pathComponents = nil
+	}
+	dataDirectoryName := filesystem.MutagenDataDirectoryName
+	if mutagen.DevelopmentModeEnabled {
+		dataDirectoryName = filesystem.MutagenDataDirectoryDevelopmentName
+	}
+	pathComponents = append(pathComponents,
+		dataDirectoryName,
+		filesystem.MutagenAgentsDirectoryName,
+		mutagen.Version,
+		BaseName,
+	)
+	return strings.Join(pathComponents, pathSeparator)
```

Comment:
Most of this could probably be moved into (another) new function like `remoteHomePath(cmdExe bool, components ...string)` so it could be used from both here and `install.go`